docs/design/milestone_1.md

```diff
@@ -0,0 +1,335 @@
+      lowest.
+    * While True:
+        * For each repository (GitHub limits this to 1000 results):
+            * If the repository has not been seen:
+                * Add it to the list of repositories
+        * If there were fewer than 1000 results:
+            * Break
+        * Set `MAX_STARS` to the the number of stars the last repository
+          returned + `STAR_OVERLAP`
+        * If `MAX_STARS` is the same as the previous value
+            * Break
+
+The current implementation of this algorithm has a difference between GitHub
+search of less than 0.05% for >=20 stars (GitHub search was checked ~12 hours
+after the algorithm finished) and took 4 hours with 1 worker and 1 token.
+
+
+##### Rate Limits
+
+A pool of GitHub tokens will be supported for increased performance.
+
+A single GitHub token has a limit of "5000" each hour, a single search page
+consumes "1", and returning the 1000 results from a search consumes "10". This
+allows 500 search queries per hour for a single token.
+
+
+##### Output
+
+Output from enumeration will be a text file containing a list of GitHub urls.
+
+
+#### Static Project URL Lists
+
+Rather than repeatedly query project repositories for a list of projects, use
+pre-generated static lists of project repository URLs.
+
+Sources:
+
+* Prior invocations of the enumeration tool
+* Manually curated lists of URLs
+* [GHTorrent](https://ghtorrent.org/) data dumps
+
+##### GHTorrent
+
+GHTorrent monitors GitHub's public event feed and provides a fairly
+comprehensive source of projects.
+
+Data from GHTorrent needs to be extracted from the SQL dump and filtered to
+eliminate deleted repositories.
+
+The 2021-03-06 dump includes approx 190M repositories. This many repositories
+would need to be curated to ensure each repository is still available. Culling
+for "interesting" (e.g. more than 1 star) repositories may also be useful to
+limit the amount of work generating signals.
+
+
+#### Future Sources of Projects
+
+There are many other sources of projects for future milestones that can be
+used. These are out-of-scope for Milestone 1, but worth listing.
+
+* Other source repositories such as GitLab and Bitbucket.
+* [https://deps.dev/](https://deps.dev/) projects. This source captures many
+  projects that exist in package repositories and helps connect projects to
+  their packages and dependents.
+* GHTorrent or GH Archive - these can avoid the expense of querying GitHub's
+  API directly.
+* Google dorking - use Google's advanced search capabilities to find
+  self-hosted repositories (e.g. cgit, gitea, etc)
+* JIRA, Bugzilla, etc support for issue tracking
+
+
+### Raw Signal Collection
+
+This stage is when the list of projects are iterated over and for each project
+a set of raw signal data is output.
+
+For Milestone 1, the focus will be on reproducing the existing signals
+collected by the Python implementation, and adding support for dependent data
+sourced from [deps.dev](https://deps.dev).
```

Comment:
Similar comment to above: this comment should be more prominent.

Reply:
Move this up to the goals section to make it more prominent.